src/main/java/pocketpal/frontend/parser/ParseCommand.java

```diff
@@ -0,0 +1,128 @@
+package pocketpal.frontend.parser;
+
+import pocketpal.frontend.commands.Command;
+import pocketpal.frontend.constants.MessageConstants;
+import pocketpal.frontend.constants.ParserConstants;
+import pocketpal.frontend.exceptions.InvalidArgumentsException;
+import pocketpal.frontend.exceptions.InvalidCategoryException;
+import pocketpal.frontend.exceptions.InvalidDateException;
+import pocketpal.frontend.exceptions.MissingArgumentsException;
+import pocketpal.frontend.exceptions.MissingDateException;
+import pocketpal.frontend.exceptions.UnknownOptionException;
+import pocketpal.frontend.util.CategoryUtil;
+import pocketpal.frontend.util.StringUtil;
+
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public abstract class ParseCommand {
+    private static final Logger logger = Logger.getLogger(ParseCommand.class.getName());
+
+    public abstract Command parseArguments(String input) throws InvalidArgumentsException,
+            InvalidCategoryException, MissingArgumentsException, MissingDateException, InvalidDateException,
+            UnknownOptionException;
+
+    public String extractDetail(String input, Pattern pattern) throws MissingArgumentsException {
+        String detail = null;
+        Matcher matcher = pattern.matcher(input);
+        if (matcher.find()) {
+            String option = matcher.group(ParserConstants.OPTION_GROUP);
+            detail = matcher.group(ParserConstants.DETAIL_GROUP);
+            if (detail != null) {
+                detail = detail.trim();
+            }
+            if ((detail == null || detail.isEmpty()) && (option != null)) { //option specified with 0 arguments
+                throw new MissingArgumentsException(MessageConstants.MESSAGE_MISSING_OPTION_ARG + option);
+            }
+        }
+        return detail;
+    }
+
+    public String extractId(String input, Pattern pattern) {
+        String id = null;
+        Matcher matcher = pattern.matcher(input);
+        if (matcher.find()) {
+            id = matcher.group(1).trim();
+        }
```

Comment:
It may be unclear what `matcher.group(1)` is referring to. Perhaps using a constant to refer to this group would be better like how `*_GROUP` is used in lines 30-31